lib/galaxy/model/store/ro_crate_utils.py

```diff
@@ -222,6 +225,216 @@ def _add_workflows(self, crate: ROCrate):
+        for step in steps:
+            if step.type == "tool":
+                tool_id = step.tool_id
+                tool_version = step.tool_version
+
+                # Cache key based on tool ID and version
+                tool_key = f"{tool_id}:{tool_version}"
+
+                # Check if tool entity is already in cache
+                if tool_key in self.tool_cache:
+                    tool_entity = self.tool_cache[tool_key]
+                else:
+                    # Create a new tool entity
+                    tool_name = tool_id
+                    tool_description = None
+                    if step.annotations:
+                        annotations_list = [annotation.annotation for annotation in step.annotations if annotation]
+                        tool_description = " ".join(annotations_list) if annotations_list else None
+
+                    # Retrieve the tool metadata from the toolbox
+                    tool_metadata = self._get_tool_metadata(tool_id)
+
+                    # Add tool entity to the RO-Crate
+                    tool_entity = crate.add(
+                        ContextEntity(
+                            crate,
+                            tool_id,
+                            properties={
+                                "@type": "SoftwareApplication",
+                                "name": tool_name,
+                                "version": tool_version,
+                                "description": tool_description,
+                                "url": "https://toolshed.g2.bx.psu.edu",  # URL if relevant
+                                "citation": tool_metadata["citations"],
+                                "identifier": tool_metadata["xrefs"],
+                                "EDAM operation": tool_metadata["edam_operations"],
+                            },
+                        )
+                    )
+
+                    # Store the tool entity in the cache
+                    self.tool_cache[tool_key] = tool_entity
+
+                # Append the tool entity to the workflow (instrument) and store it in the list
+                tool_entities.append(tool_entity)
+                crate.mainEntity.append_to("instrument", tool_entity)
+
+            # Handle subworkflows recursively
+            elif step.type == "subworkflow":
+                subworkflow = step.subworkflow
+                if subworkflow:
+                    self._add_tools_recursive(subworkflow.steps, crate, tool_entities)
+
+    def _get_tool_metadata(self, tool_id: str):
+        """
+        Retrieve the tool metadata (citations, xrefs, EDAM operations) using the ToolBox.
+
+        Args:
+            toolbox (ToolBox): An instance of the Galaxy ToolBox.
+            tool_id (str): The ID of the tool to retrieve metadata for.
+
+        Returns:
+            dict: A dictionary containing citations, xrefs, and EDAM operations for the tool.
+        """
+        tool = self.toolbox.get_tool(tool_id)
```

Comment:
This won't work, you can't use the toolbox in this context. We can't load up the toolbox in celery workers, that would take too much time and memory. The best solution is to either get that data from a yet to be written toolshed endpoint or load up the tool on demand, neither of these are easy to do or fit into the context of this PR. Maybe you could focus on the other enhancements in this PR.

Reply:
That's a shame that the toolbox can't be used in this way. But the most essential metadata about the tools is already captured without using the toolbox, so I think it's ok to remove this function and consider this as an enhancement which could be added in future.

It does mean we would have a greater need for the `url` on line 369, though, as without the `xrefs` there's no alternative links to find the tool (and the rest of its metadata). Still not absolutely mandatory to have `url`, but a definite best practice.

(I wondered if we could retrieve the citation/xref/EDAM metadata from something like `job_attrs.txt` which is already part of the crate, but it's not included there. Nor is the url.)